File: Data Structures & Algorithms/course-schedule-ii/submission-2.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        n = numCourses 
        graph = defaultdict(list) 
        indegree = [ 0 for i in range(n) ] 

        for after, must in prerequisites: 
            indegree[after] += 1 
            graph[must].append(after) 

        
        q = deque() 

        for i in range(n): 
            if indegree[i] == 0: 
                q.append(i) 

        count = 0 
        res = []
        while q: 
            cur = q.popleft() 
            count += 1 
            res.append(cur)

            for nei in graph[cur]: 
                indegree[nei] -= 1 
                if indegree[nei] == 0: 
                    q.append(nei) 

        return res if count == n else []

File: Data Structures & Algorithms/course-schedule/submission-2.js
class Solution {
    /**
     * @param {number} numCourses
     * @param {number[][]} prerequisites
     * @return {boolean}
     */
    canFinish(numCourses, prerequisites) { 
        const indegree = new Array(numCourses).fill(0) 
        const graph = new Array(numCourses).fill(null).map(() => []) 

        for (const [after, must] of prerequisites) {
            graph[must].push(after) 
            indegree[after] ++
        } 

        let q = [] 
        for (let i = 0; i < indegree.length; i ++) {
            if (indegree[i] === 0) {
                q.push(i)
            }
        } 

        let head = 0 
        let count = 0 
        while(head < q.length) { 
            let course = q[head ++]
            count ++ 
            for (const nei of graph[course]) {
                indegree[nei] -- 
                if (indegree[nei] === 0) {
                    q.push(nei)
                }
            }
        } 

        return count === numCourses
    }
}
